src/ann_solo/utils.py

```diff
@@ -1,104 +1,457 @@
+        )
+        features["abs_mz_diff_da"].append(
+            abs(
+                mass_diff(
+                    ssm.query_spectrum.precursor_mz,
+                    ssm.library_spectrum.precursor_mz,
+                    True,
+                )
+            )
+        )
+        features["cosine"].append(sim_calc.cosine())
+        features["cosine_top5"].append(sim_calc_top.cosine())
+        features["n_matched_peaks"].append(sim_calc.n_matched_peaks())
+        features["frac_n_peaks_query"].append(sim_calc.frac_n_peaks_query())
+        features["frac_n_peaks_lib"].append(sim_calc.frac_n_peaks_library())
+        features["frac_n_peaks_lib_top5"].append(
+            sim_calc_top.frac_n_peaks_library()
+        )
+        features["frac_int_query"].append(sim_calc.frac_intensity_query())
+        features["frac_int_lib"].append(sim_calc.frac_intensity_library())
+        features["frac_int_lib_top5"].append(
+            sim_calc_top.frac_intensity_library()
+        )
+        features["mse_mz"].append(sim_calc.mean_squared_error("mz"))
+        features["mse_mz_top5"].append(sim_calc_top.mean_squared_error("mz"))
+        features["mse_int"].append(sim_calc.mean_squared_error("intensity"))
+        features["mse_int_top5"].append(
+            sim_calc_top.mean_squared_error("intensity")
+        )
+        features["contrast_angle"].append(sim_calc.spectral_contrast_angle())
+        features["contrast_angle_top5"].append(
+            sim_calc_top.spectral_contrast_angle()
+        )
+        features["hypergeometric_score"].append(
+            sim_calc.hypergeometric_score(
+                min_mz=config.min_mz,
+                max_mz=config.max_mz,
+                fragment_mz_tol=config.bin_size,
+            )
+        )
+        features["kendalltau"].append(sim_calc.kendalltau())
+        features["ms_for_id_v1"].append(sim_calc.ms_for_id_v1())
+        features["ms_for_id_v2"].append(sim_calc.ms_for_id_v2())
+        features["entropy_unweighted"].append(sim_calc.entropy(False))
+        features["entropy_weighted"].append(sim_calc.entropy(True))
+        features["scribe_fragment_acc"].append(sim_calc.scribe_fragment_acc())
+        features["scribe_fragment_acc_top5"].append(
+            sim_calc_top.scribe_fragment_acc()
+        )
+        features["manhattan"].append(sim_calc.manhattan())
+        features["euclidean"].append(sim_calc.euclidean())
+        features["chebyshev"].append(sim_calc.chebyshev())
+        features["pearsonr"].append(sim_calc.pearsonr())
+        features["pearsonr_top5"].append(sim_calc_top.pearsonr())
+        features["spearmanr"].append(sim_calc.spearmanr())
+        features["spearmanr_top5"].append(sim_calc_top.spearmanr())
+        features["braycurtis"].append(sim_calc.braycurtis())
+        features["canberra"].append(sim_calc.canberra())
+        features["ruzicka"].append(sim_calc.ruzicka())
+        features["is_target"].append(not ssm.is_decoy)
+    return features
```

Comment:
**question:** Is performance not a problem here?

Reply:
Agree. This can and should be optimized, i.e. by multithreading and by using Numba. My main focus was getting the full functionality implemented first, but we can probably think about performance now as well. Multithreading should normally be a trivial change.